Plugins/NativeXr/Source/NativeXr.cpp

```diff
@@ -468,18 +476,27 @@ namespace Babylon
                             m_sessionState->Session = std::move(session);
                             NotifySessionStateChanged(true);
                         });
-                });
+                }));
+
+            return m_beginTask.value();
         }
 
         arcana::task<void, std::exception_ptr> NativeXr::Impl::EndSessionAsync()
         {
+            assert(m_beginTask);
+            assert(m_sessionState != nullptr);
+
             m_sessionState->CancellationSource.cancel();
 
             m_sessionState->FrameBufferToJsTextureMap.clear();
             m_sessionState->TextureToFrameBufferMap.clear();
             m_sessionState->ActiveTextures.clear();
 
-            return m_sessionState->FrameTask.then(m_sessionState->GraphicsImpl.AfterRenderScheduler(), arcana::cancellation::none(), [this, thisRef{shared_from_this()}](const arcana::expected<void, std::exception_ptr>&) {
+            // Don't try to end the session while it is still starting.
+            m_endTask = m_beginTask->then(arcana::inline_scheduler, arcana::cancellation::none(), [this, thisRef{shared_from_this()}] {
```

Comment:
when_all?

Reply:
I considered it, but I feel like the logic is more clear as is. If you feel strongly I can change it, but in a separate PR since I already tested it as is in multiple apps.